### ISSUE TITLE
Error Rates off 100 fold

### ISSUE THREAD

**Issue by ShayneWierbowski** [sej917/BEAR#12]
To whom it may concern,

I am working on a project generating reads using BEAR and I think I have found an error in the code that causes the rates of substitutions and indels to be effectively 100x what they should be. I could be vastly mistaken in my understanding of the code, but the code is setup as follows...

1) Creating the error profile using DRISEE (step 2 in the user manual) produces error.per which from what I can tell contains the probability that a substitution at A C G or T or an indel (X) occurs. These values are stored as a **PERCENTAGE from 0 to 100**.

2) Then, when the error profiles are applied in trim_reads.py (step 4.b in the user manual) the error.per file is read in by error_models.py to produce a regression curve to predict error rates at each bp position. However, later on the rand() function is used to determine whether or not each of these errors should be applied (ie if rand() < error_rate, apply the error). Since **rand() returns a float between 0.0 and 1.0** this will almost always be the case, especially towards the end of reads.

For example, a substitution that should have occurred with a rate of 1% will always occur since rand() is always less than 1. In order to resolve this issue, error_models.py must first convert values from a percentage (0-100) to a probability (0.0-1.0).

If I am correct about this, then this is a grievous error and quite frankly, I am surprised it made it through your testing, particularly since your article in BMC Bioinformatics seems to be calculating the rates correctly in the plots. Nonetheless, a commit fixing this error should be added ASAP.

**Comment by beacamara** [comment]
Dear ShayneWierbowski,

I am trying to use BEAR to generate metagenomic reads, and I have just realized of your message. Did you try to convert values from percentage to probability before the rand() function?

Thank you so much for your help

Beatriz

**Comment by ShayneWierbowski** [comment]
Beatriz,

I imagine there are multiple ways you could go about fixing this issue. What I found was the easiest solution for my needs, was to modify the file called error_models.py. When you call trim_reads.pl, it reads in the substitution rates from error.per (which are stored on a 0-100 scale) by executing error_models.py. So if you fix error_models.py to return probabilities, then you fix the problem early without having to play around with all of the comparisons to the rand() values generated in the rest of trim_reads.pl.

So the solution that I have was to change line 6 in error_models.py from...

d = np.loadtxt(sys.argv[1], skiprows=3)

to...

d = np.loadtxt(sys.argv[1], skiprows=3) / 100.0

I don't actually have the original source error_models.py anymore so I am somewhat guessing that that is what the line had previously. I think that should fix the error rates for you, but by all means you should play around with it to make sure that you are seeing the same problems I saw, and that my suggested fix actually does fix it.

Hope that helps.

**Comment by beacamara** [comment]
Dear Dhayne Wierbowski,

Thank you so much for your help. I introduced the modifications you suggested me in  the error_models.py and it worked perfectly. However, there was another issue to report. We have realized that trim_reads.pl program did not generate the reverse complement reads for paired-end reads because a mistake in the program. My workmate has already reported this fact in #14. I was wondering if you noticed this.

Again thank you so much

Bea
